src/resolve/resolved-def.h

```diff
@@ -83,4 +83,38 @@
 
 #define SD_RESOLVED_FROM_MASK       (SD_RESOLVED_FROM_CACHE|SD_RESOLVED_FROM_ZONE|SD_RESOLVED_FROM_TRUST_ANCHOR|SD_RESOLVED_FROM_NETWORK)
 
+/* LLMNR Jitter interval, see RFC 4795 Section 7 */
+#define SD_RESOLVED_LLMNR_JITTER_INTERVAL_USEC (100 * USEC_PER_MSEC)
+
+/* mDNS probing interval, see RFC 6762 Section 8.1 */
+#define SD_RESOLVED_MDNS_PROBING_INTERVAL_USEC (250 * USEC_PER_MSEC)
+
+/* Maximum attempts to send DNS requests, across all DNS servers */
+#define SD_RESOLVED_DNS_TRANSACTION_ATTEMPTS_MAX 24
+
+/* Maximum attempts to send LLMNR requests, see RFC 4795 Section 2.7 */
+#define SD_RESOLVED_LLMNR_TRANSACTION_ATTEMPTS_MAX 3
+
+/* Maximum attempts to send MDNS requests, see RFC 6762 Section 8.1 */
+#define SD_RESOLVED_MDNS_TRANSACTION_ATTEMPTS_MAX 3
+
+#define SD_RESOLVED_TRANSACTION_ATTEMPTS_MAX(p) (\
+                                     (p) == DNS_PROTOCOL_LLMNR ?        \
+                                     SD_RESOLVED_LLMNR_TRANSACTION_ATTEMPTS_MAX :   \
+                                     (p) == DNS_PROTOCOL_MDNS ?         \
+                                     SD_RESOLVED_MDNS_TRANSACTION_ATTEMPTS_MAX :    \
+                                     SD_RESOLVED_DNS_TRANSACTION_ATTEMPTS_MAX)
+
 #define SD_RESOLVED_QUERY_TIMEOUT_USEC (120 * USEC_PER_SEC)
+
+/* After how much time to repeat classic DNS requests */
+#define SD_RESOLVED_UDP_TIMEOUT_USEC (SD_RESOLVED_QUERY_TIMEOUT_USEC / SD_RESOLVED_DNS_TRANSACTION_ATTEMPTS_MAX)
+
+/* When we do TCP, grant a much longer timeout, as in this case there's no need for us to quickly
+ * resend, as the kernel does that anyway for us, and we really don't want to interrupt it in that
+ * needlessly. */
+#define SD_RESOLVED_TCP_TIMEOUT_USEC (10 * USEC_PER_SEC)
+
+/* Should be longer than transaction timeout for a single UDP transaction, so we get at least
+ * one transaction retry before timeouting the whole candidate */
+#define SD_RESOLVED_CANDIDATE_EXPEDITED_TIMEOUT_USEC (SD_RESOLVED_UDP_TIMEOUT_USEC + 1*USEC_PER_SEC)
```

Comment:
oh, please drop the SD_ prefix again, we only use that for definitions we consider exported, in the sense that other packages (i.e. not systemd) might use them. The definitions you see above in this file are necessary if you want to interface with resolved's dbus interface, hence they carry the SD_ prefix. (We do not *actually* export them via this header file, admittedly, but we do carry a copy of the relevant defines in the org.freedesktop.resolve1(5) man page.)

I think these timeouts really shouldn't be made public, we want the liberty to tweak them without this being considered an api break. Hence these defines really shouldn't have the SD_ prefix, and probably also not be in resolved-def.h.

If you want, add a new header file resolved-timeouts.h or so with these defines (without the SD_ prefix then), but maybe leave resolved-def.h as is, only with the "public" defines.

Reply:
That makes sense to me, I'll do that.

And I guess `SD_RESOLVED_QUERY_TIMEOUT_USEC` sort-of is a special case? It's not mentioned in the manpage, but it lives in the header so `nss-resolve.c` can use it?

Reply:
yeah, we probably should document it in the man page. the default method timeout most dbus libs apply is too short for DNS queries, hence we have the define, and it must be applied client side.